File: scripts/fetch_stocks.js
#!/usr/bin/env node

const tickers = process.argv.slice(2);
if (!tickers.length) {
  console.error("Usage: fetch_stocks.js TICKER1 TICKER2 ...");
  process.exit(1);
}

const nyseURL = "https://www.nyse.com/api/nyseservice/v1/quotes?symbol=";
const csURL = "https://www.schwab.wallst.com/Prospect/Research/mutualfunds/fees.asp?symbol=";
const reTABLE = /^[\s\S]*(<table id="firstGlanceQuoteTable".*?<\/table>)[\s\S]*$/m;
const rePRICE = /^.*?<tbody><tr><td>\$([0-9\.]+)<\/td>.*$/;
const reCHANGE = /^.*?<tbody><tr><td>\$[0-9\.]+<\/td><td><span.*?>([+-]?[0-9\.]+)<\/span>.*$/;

async function fetchNyse(ticker) {
  const response = await fetch(nyseURL + ticker, {
    headers: {
      "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
      "accept-language": "en-US,en;q=0.9",
      "cache-control": "max-age=0",
      "sec-ch-ua": "\"Chromium\";v=\"143\", \"Not A(Brand\";v=\"24\"",
      "sec-ch-ua-mobile": "?0",
      "sec-ch-ua-platform": "\"macOS\"",
      "sec-fetch-dest": "document",
      "sec-fetch-mode": "navigate",
      "sec-fetch-site": "none",
      "sec-fetch-user": "?1",
      "upgrade-insecure-requests": "1"
    },
  });
  const nyse = await response.json();
  const price = nyse?.quote?.last;
  const change = nyse?.quote?.change;
  if (/^[0-9.]+$/.test(String(price))) {
    return { price: parseFloat(price), change: parseFloat(change) || 0 };
  }
  return null;
}

async function fetchSchwab(ticker) {
  const response = await fetch(csURL + ticker);
  const html = await response.text();
  const table = html.replace(reTABLE, "$1").replace(/[\r\n]/g, "");
  const price = table.replace(rePRICE, "$1");
  const change = table.replace(reCHANGE, "$1");
  if (/^[0-9.]+$/.test(price)) {
    return { price: parseFloat(price), change: parseFloat(change) || 0 };
  }
  return null;
}

async function fetchTicker(ticker) {
  try {
    const result = await fetchNyse(ticker);
    if (result) return result;
  } catch (e) { /* fall through */ }
  try {
    const result = await fetchSchwab(ticker);
    if (result) return result;
  } catch (e) { /* fall through */ }
  return null;
}

async function main() {
  const results = {};
  const promises = tickers.map(async (ticker) => {
    const data = await fetchTicker(ticker);
    if (data) results[ticker] = data;
  });
  await Promise.all(promises);
  console.log(JSON.stringify({ data: results }));
}

main().catch((err) => {
  console.error(err.message);
  process.exit(1);
});

File: scripts/fetch_stocks.py
#!/usr/bin/env python3
"""Fetch stock prices from NYSE API with Schwab HTML fallback."""

import json
import re
import sys
import urllib.request
from concurrent.futures import ThreadPoolExecutor

NYSE_URL = "https://www.nyse.com/api/nyseservice/v1/quotes?symbol="
CS_URL = "https://www.schwab.wallst.com/Prospect/Research/mutualfunds/fees.asp?symbol="

RE_TABLE = re.compile(r"^[\s\S]*(<table id=\"firstGlanceQuoteTable\".*?</table>)[\s\S]*$", re.DOTALL)
RE_PRICE = re.compile(r"^.*?<tbody><tr><td>\$([0-9.]+)</td>.*$")
RE_CHANGE = re.compile(r"^.*?<tbody><tr><td>\$[0-9.]+</td><td><span.*?>([+-]?[0-9.]+)</span>.*$")


def fetch_nyse(ticker):
    req = urllib.request.Request(
        NYSE_URL + ticker,
        headers={
            "Accept": "application/json",
            "User-Agent": "Mozilla/5.0",
        },
    )
    with urllib.request.urlopen(req, timeout=10) as resp:
        data = json.loads(resp.read())
    price = data.get("quote", {}).get("last")
    change = data.get("quote", {}).get("change")
    if price is not None and re.match(r"^[0-9.]+$", str(price)):
        return {"price": float(price), "change": float(change or 0)}
    return None


def fetch_schwab(ticker):
    req = urllib.request.Request(CS_URL + ticker, headers={"User-Agent": "Mozilla/5.0"})
    with urllib.request.urlopen(req, timeout=10) as resp:
        html = resp.read().decode("utf-8", errors="replace")
    table = RE_TABLE.sub(r"\1", html).replace("\r", "").replace("\n", "")
    price = RE_PRICE.sub(r"\1", table)
    change = RE_CHANGE.sub(r"\1", table)
    if re.match(r"^[0-9.]+$", price):
        return {"price": float(price), "change": float(change) if re.match(r"^[+-]?[0-9.]+$", change) else 0}
    return None


def fetch_ticker(ticker):
    for fn in (fetch_nyse, fetch_schwab):
        try:
            result = fn(ticker)
            if result:
                return result
        except Exception:
            pass
    return None


def main():
    tickers = sys.argv[1:]
    if not tickers:
        print("Usage: fetch_stocks.py TICKER1 TICKER2 ...", file=sys.stderr)
        sys.exit(1)

    results = {}
    with ThreadPoolExecutor(max_workers=10) as pool:
        futures = {pool.submit(fetch_ticker, t): t for t in tickers}
        for future in futures:
            ticker = futures[future]
            try:
                data = future.result()
                if data:
                    results[ticker] = data
            except Exception:
                pass

    print(json.dumps({"data": results}))


if __name__ == "__main__":
    main()
